### PR TITLE
feat: propagate shadowroot attributes

### PR DESCRIPTION
**Pull request by janechu**
# Pull Request

## 📖 Description

- Propagates `shadowroot*` attributes from `f-template` onto rendered declarative shadow DOM `<template>` elements.
- Preserves the existing default `shadowrootmode="open"` and `shadowroot="open"` behavior when no mode is provided.
- Updates FAST build configuration/docs and adds Rust/Node coverage for explicit, boolean, escaped, and extra shadowroot attributes.

## 👩‍💻 Reviewer Notes

Please focus review on the shadowroot attribute normalization/escaping path shared between the Rust renderer and the JS CLI configuration bridge.

## 📑 Test Plan

Passed:

- `git diff --check`
- `cargo test --manifest-path crates/microsoft-fast-build/Cargo.toml`
- `npm run build -w @microsoft/fast-build`
- `npm run test:node -w @microsoft/fast-build`
- `npm run build`
- `npm run biome:check`
- `npm run checkchange`

### General

- [ ] I have included a change request file using `$ npm run change`
- [x] I have added tests for my changes.
- [x] I have tested my changes.
- [x] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/microsoft/fast/blob/main/CONTRIBUTING.md) documentation and followed the [standards](https://github.com/microsoft/fast/blob/main/CODE_OF_CONDUCT.md#our-standards) for this project.

### CODE REVIEW

File: crates/microsoft-fast-build/src/directive.rs [hunk 1]
```diff
@@ -371,13 +372,77 @@ pub fn render_custom_element(
...
+        }
+    }
+
+    let mode = mode
+        .and_then(|value| value)
+        .unwrap_or_else(|| "open".to_string());
+    let legacy_shadowroot = match legacy_shadowroot {
+        Some(Some(value)) => value,
+        Some(None) | None => mode.clone(),
+    };
```
Comment (janechu) [inline crates/microsoft-fast-build/src/directive.rs]: **Asymmetric mirroring between `shadowrootmode` and `shadowroot`.** The mirror only goes `mode → legacy`, not `legacy → mode`. If a user writes `<f-template shadowroot="closed">` (without `shadowrootmode`), this code produces:

```
<template shadowrootmode="open" shadowroot="closed">
```

Modern browsers (which honor `shadowrootmode`) will mount the shadow as **open**, while legacy browsers (which honor the old `shadowroot` attribute) see **closed**. That cross-browser inconsistency is unlikely to be what the author intended.

The README states *"When `shadowrootmode` is provided, its value is also used for the legacy `shadowroot` attribute unless `shadowroot` is explicitly provided"* — but it doesn't address the reverse, and currently an explicitly-provided legacy `shadowroot="closed"` is silently ignored by the modern-mode default.

Suggestion: when `mode` resolves to its `"open"` default *because no explicit `shadowrootmode` value was given* (i.e. it was absent or boolean) and `legacy_shadowroot` has an explicit value, mirror the legacy value into mode as well — so the two attributes always agree when only one was explicitly set. Either way, please add a test that pins down the chosen semantics for `shadowroot="closed"` alone.

File: crates/microsoft-fast-build/src/directive.rs [hunk 2]
```diff
@@ -371,13 +372,77 @@ pub fn render_custom_element(
...
+                    forwarded.push((normalized, value.clone()));
+                }
+            }
+            _ => {}
+        }
+    }
+
+    let mode = mode
+        .and_then(|value| value)
+        .unwrap_or_else(|| "open".to_string());
```
Comment (janechu) [inline crates/microsoft-fast-build/src/directive.rs:416]: **Boolean vs empty-string `shadowrootmode` produce different output.**

- `<f-template shadowrootmode>` (boolean) → `shadowrootmode="open"` (default applied)
- `<f-template shadowrootmode="">` (explicit empty) → `shadowrootmode=""` (passes through; browsers will not mount a shadow root)

Reached this by tracing `mode.and_then(|value| value)`: `Some(None)` flattens to `None` (default applied) but `Some(Some(""))` stays as `Some("")`. The two HTML forms are arguably equivalent (HTML5 says the empty-string value should be treated like the boolean form), so it'd be more consistent to also fall back to `"open"` when the value is the empty string. Same applies to the `shadowroot` legacy attribute below. If pass-through of empty values is intentional, a quick comment here would help future maintainers.

File: crates/microsoft-fast-build/src/locator.rs [hunk 3]
```diff
@@ -66,27 +77,53 @@ impl Locator {
...
         Ok(Locator { templates })
     }
 
     /// Create a Locator from an explicit map (useful for testing without filesystem).
     pub fn from_templates(templates: HashMap<String, String>) -> Self {
+        let templates = templates
+            .into_iter()
+            .map(|(name, content)| (name, TemplateDefinition { content, shadowroot_attributes: Vec::new() }))
+            .collect();
+        Locator { templates }
```
Comment (janechu) [inline crates/microsoft-fast-build/src/locator.rs:103]: **Public Rust API gap.** `from_templates` (and `add_template` below) silently drop shadowroot-attribute support — programmatic Rust callers can't set them at all without going through `from_patterns` or the WASM bridge. `from_template_definitions` exists but is `pub(crate)`, so it's not an option for downstream crates either.

Suggest one of:
1. Promote `from_template_definitions` to `pub` (with a stable parameter type — a struct rather than a tuple), or
2. Add `add_template_with_shadowroot_attrs(&mut self, name: &str, content: &str, attrs: &[(String, Option<String>)])`, or
3. Document on `from_templates`/`add_template` that they intentionally do not support shadowroot attributes and point users to `<f-template>`.

At minimum option 3 is needed so the asymmetry doesn't surprise consumers.

File: crates/microsoft-fast-build/tests/custom_elements.rs [hunk 4]
```diff
@@ -90,6 +90,108 @@ fn test_custom_element_with_children() {
...
+    let locator = Locator::from_patterns(&["tests/fixtures/shadowroot-explicit.html"]).unwrap();
+    let result = render_with_locator(
+        r#"<shadowroot-explicit></shadowroot-explicit>"#,
+        &empty_root(),
+        &locator,
+        None,
+    ).unwrap();
+    assert!(result.contains(r#"<template shadowrootmode="closed" shadowroot="open">"#), "shadow attrs: {result}");
+}
+
```
Comment (janechu) [inline crates/microsoft-fast-build/tests/custom_elements.rs:155]: Test coverage gap: there is no fixture/case for **`shadowroot="closed"` alone (no `shadowrootmode`)**, which is exactly the edge case where mode defaults to `"open"` while legacy is `"closed"`. Whatever direction the asymmetric-mirror conversation goes (see comment in `directive.rs`), please add a test that pins the behavior down. Suggested fixture name `shadowroot-legacy-only.html` with `<f-template ... shadowroot="closed">`.

**Review by janechu — COMMENTED** [review]
The parsing/escaping path is solid and the test coverage for the common cases is good. A few concerns left as inline comments. The most notable is an asymmetric `shadowrootmode` ↔ `shadowroot` mirroring: when a user specifies only `shadowroot="closed"` (no `shadowrootmode`), the output is `<template shadowrootmode="open" shadowroot="closed">`, which gives modern browsers "open" and legacy ones "closed". I verified this by running the new logic against that input. Other comments cover a boolean-vs-empty-value distinction, a public Rust API gap, missing test coverage for the asymmetric case, and a small JSON parsing nit.

File: crates/microsoft-fast-build/src/wasm.rs [hunk 5]
```diff
@@ -139,6 +161,73 @@ fn parse_templates_map(templates_json: &str) -> Result<HashMap<String, String>,
...
+
+    let attrs = match template.get("shadowrootAttributes") {
+        Some(json::JsonValue::Array(attrs)) => parse_shadowroot_attributes(template_name, attrs)?,
+        Some(_) => {
+            return Err(JsValue::from_str(&format!(
+                "Template metadata for '{}' must use an array shadowrootAttributes property",
+                template_name
+            )));
+        }
+        None => Vec::new(),
```
Comment (janechu) [inline crates/microsoft-fast-build/src/wasm.rs]: Minor: a JSON `"shadowrootAttributes": null` would fall into the `Some(_)` arm above and surface as `"must use an array shadowrootAttributes property"`. JSON consumers commonly emit `null` instead of an empty array (some serializers do this by default). Consider treating `JsonValue::Null` the same as a missing property to be more forgiving:

```rust
let attrs = match template.get("shadowrootAttributes") {
    Some(json::JsonValue::Array(attrs)) => parse_shadowroot_attributes(template_name, attrs)?,
    Some(json::JsonValue::Null) | None => Vec::new(),
    Some(_) => { /* error */ }
};
```

File: packages/fast-build/test/config.test.js [hunk 6]
```diff
@@ -208,6 +208,52 @@ describe("backward compatibility", () => {
...
+                "--state=state.json",
+                "--output=out.html",
+            ],
+            dir,
+        );
+
+        const output = fs.readFileSync(path.join(dir, "out.html"), "utf8");
+        assert.ok(
+            output.includes(
+                `<template shadowrootmode="closed" shadowroot="closed" shadowrootdelegatesfocus>`,
```
Comment (janechu) [inline packages/fast-build/test/config.test.js]: Nit: this assertion is order-sensitive (`shadowrootmode` then `shadowroot` then forwarded). That mirrors the current Rust ordering, but it tightly couples the JS-CLI test to a Rust implementation detail. Consider asserting attribute presence individually (e.g. three `output.includes(...)` checks) so a future change to the order of forwarded attributes doesn't break this test unrelated to a real regression.

**Review by janechu — COMMENTED** [review]
Re-review of `e674d99` — all six previous comments are addressed cleanly:

1. **Asymmetric mirror** ✅ — new `explicit_shadowroot_attr_value` helper plus the `or_else`/`or` chain mirrors `shadowroot → shadowrootmode` as well, so `<f-template shadowroot="closed">` now yields `shadowrootmode="closed" shadowroot="closed"`. New test `test_f_template_legacy_shadowroot_value_mirrors_to_mode` + fixture `shadowroot-legacy-closed.html` pin it down.
2. **Boolean vs empty-string** ✅ — `explicit_shadowroot_attr_value` filters empty strings out, so `shadowrootmode=""` is now treated identically to the boolean form. Four new fixtures + tests cover the empty cases (mode-only, shadowroot-only, mode-empty + shadowroot="closed", mode="closed" + shadowroot-empty).
3. **Public Rust API gap** ✅ — added `add_template_with_shadowroot_attrs` plus rustdoc on `from_templates` and `add_template` pointing users there. Unit test `test_add_template_with_shadowroot_attrs_preserves_attrs` added.
4. **Test coverage gap** ✅ — covered by item #1.
5. **JSON `null` shadowrootAttributes** ✅ — `Some(json::JsonValue::Null) | None => Vec::new()` arm added; wasm-test added (note: gated to `target_arch = "wasm32"`, which matches the rest of `wasm.rs`).
6. **Order-coupled JS assertion** ✅ — refactored to extract the `<template …>` tag once and assert each attribute's presence individually.

Verified with `cargo test` (65 unit + 60 integration tests pass) and `npm run test:node -w @microsoft/fast-build` (30/30 pass). Also re-ran the truth table for `build_shadowroot_template_attrs` — explicit conflicts like `shadowrootmode="open" shadowroot="closed"` are still preserved verbatim, which is the right call.

One small follow-up left as an inline comment: the user-facing docs still only describe the one-way `mode → shadowroot` mirror.

File: crates/microsoft-fast-build/README.md [hunk 7]
```diff
@@ -318,6 +318,8 @@ A single file may contain multiple templates:
 
 If an `<f-template>` element is missing a `name` attribute, a warning is emitted to stderr and the template is ignored.
 
+Any `shadowroot`-prefixed attributes on `<f-template>` are copied to the rendered Declarative Shadow DOM `<template>`. When `shadowrootmode` is omitted, output continues to default to `shadowrootmode="open"` and `shadowroot="open"`. When `shadowrootmode` is provided, its value is also used for the legacy `shadowroot` attribute unless `shadowroot` is explicitly provided.
```
Comment (janechu) [inline crates/microsoft-fast-build/README.md]: Doc lag: this paragraph (and the matching ones in `crates/microsoft-fast-build/DESIGN.md`, `packages/fast-build/README.md`, and `packages/fast-build/DESIGN.md`) still describes only the one-way `shadowrootmode → shadowroot` mirror, but the implementation now also:

1. Mirrors `shadowroot → shadowrootmode` when mode is absent or empty (e.g. `shadowroot="closed"` alone now produces `shadowrootmode="closed" shadowroot="closed"`).
2. Treats an empty-string value (`shadowrootmode=""` / `shadowroot=""`) the same as a boolean attribute and falls back to defaults / the other attribute's value.

Worth documenting since both behaviors can otherwise be surprising. Suggested rewording for this line:

> Any `shadowroot`-prefixed attributes on `<f-template>` are copied to the rendered Declarative Shadow DOM `<template>`. When neither `shadowrootmode` nor `shadowroot` is set (or both are empty / boolean), output defaults to `shadowrootmode="open" shadowroot="open"`. When only one is set with a value, that value is mirrored to the other so modern and legacy browsers agree. Explicit conflicting values (e.g. `shadowrootmode="open" shadowroot="closed"`) are preserved as-is.

**Review by janechu — COMMENTED** [review]
Re-review of `0307207` — doc lag is fully addressed across all four files (`crates/microsoft-fast-build/{DESIGN,README}.md`, `packages/fast-build/{DESIGN,README}.md`). The new wording is consistent across locations and accurately describes the implemented behavior:

- Neither attribute non-empty → defaults to `shadowrootmode="open" shadowroot="open"`
- Exactly one has a non-empty value → mirrored to the other
- Both have explicit non-empty values → preserved as authored

Cross-checked against the truth table of `build_shadowroot_template_attrs`/`explicit_shadowroot_attr_value` — all four cases match.

No further blocking issues.

**Review by chrisdholt — APPROVED** [review]
